fpga_interchange/xdc.cc

```diff
@@ -0,0 +1,160 @@
+    PortInfo * port_info = static_cast<PortInfo*>(objPtr->internalRep.twoPtrValue.ptr2);
+
+    std::string port_name = port_info->name.str(ctx);
+    set_tcl_obj_string(objPtr, port_name);
+
+}
+
+static void port_dup(Tcl_Obj *srcPtr, Tcl_Obj *dupPtr) {
+    dupPtr->internalRep.twoPtrValue = srcPtr->internalRep.twoPtrValue;
+}
+
+static void port_free(Tcl_Obj *objPtr) {
+}
+
+
+static void Tcl_SetStringResult(Tcl_Interp *interp, const std::string &s) {
+    char * copy = Tcl_Alloc(s.size()+1);
+    std::copy(s.begin(), s.end(), copy);
+    copy[s.size()] = '\0';
+    Tcl_SetResult(interp, copy, TCL_DYNAMIC);
+}
+
+
+static Tcl_ObjType port_object = {
+    "port",
+    port_free,
+    port_dup,
+    port_update_string,
+    port_set_from_any,
+};
+
+static int get_ports(
+        ClientData data,
+        Tcl_Interp *interp,
+        int objc,
+        Tcl_Obj *CONST objv[]) {
+    const Context *ctx = static_cast<const Context*>(data);
+    if(objc == 1) {
+        // Return list of all ports.
+        Tcl_SetStringResult(interp, "Unimplemented");
+        return TCL_ERROR;
+    } else if(objc == 2) {
+        const char * arg0 = Tcl_GetString(objv[1]);
+        IdString port_name = ctx->id(arg0);
+
+        auto iter = ctx->ports.find(port_name);
+        if(iter == ctx->ports.end()) {
+            Tcl_SetStringResult(interp, "Could not find port " + port_name.str(ctx));
+            return TCL_ERROR;
+        }
+
+        Tcl_Obj * result = Tcl_NewObj();
+        result->typePtr = &port_object;
+        result->internalRep.twoPtrValue.ptr1 = (void*)(ctx);
+        result->internalRep.twoPtrValue.ptr2 = (void*)(&iter->second);
+
+        result->bytes = nullptr;
+        port_update_string(result);
+
+        Tcl_SetObjResult(interp, result);
+        return TCL_OK;
+    } else {
+        return TCL_ERROR;
+    }
+}
+
+static int set_property(
+        ClientData data,
+        Tcl_Interp *interp,
+        int objc,
+        Tcl_Obj *CONST objv[]) {
+    // set_property <property> <value> <object>
+    if(objc != 4) {
+        Tcl_SetStringResult(interp, "Only simple 'set_property <property> <value> <object>' is supported");
+        return TCL_ERROR;
+    }
+
+    const char *property = Tcl_GetString(objv[1]);
+    const char *value = Tcl_GetString(objv[2]);
+    const Tcl_Obj *object = objv[3];
+
+    if(object->typePtr != &port_object) {
+        Tcl_SetStringResult(interp, "Only port objects are handled right now!");
+        return TCL_ERROR;
+    }
+
+    const Context *ctx = static_cast<const Context*>(object->internalRep.twoPtrValue.ptr1);
+    PortInfo * port_info = static_cast<PortInfo*>(object->internalRep.twoPtrValue.ptr2);
+    NPNR_ASSERT(port_info->net != nullptr);
+    CellInfo * cell = ctx->port_cells.at(port_info->name);
+
+    cell->attrs[ctx->id(property)] = Property(value);
+
+    return TCL_OK;
+}
+
+void Arch::parse_xdc(const std::string &filename) {
```

Comment:
As a bit of a note; a proper xdc implementation will need to override the default command handler to return "number" commands as themselves in square brackets; so that bus ports work without needing an escaped string (e.g. command `3` returns `[3]` so you can do `foo[3]`). It turns out the rules Vivado use for this are actually [quite interesting](https://twitter.com/whitequark/status/1256508818554597377)...

Reply:
Good catch, I'll fix that.

Reply:
This was addressed in #582